File: Problems/209/main.go
package main

func main() {}

func minSubArrayLen(target int, nums []int) int {
	// Naive solution: slide throught whole list
	// Time complexity: O(nlogn), space complexity: O(1)
	if len(nums) == 0 {
		return 0
	}

	res := 100001
	st, ed, sum := 0, 1, nums[0]

	// Sliding windows
	for st < len(nums) && !(ed == len(nums) && sum < target) {
		for sum < target && ed < len(nums) {
			sum += nums[ed]
			ed++
		}

		tmp := ed - st
		if sum >= target && tmp < res {
			res = tmp
		}
		sum -= nums[st]
		st++
	}

	if res == 100001 {
		return 0
	}
	return res
}

File: Problems/209/main.cpp
#include <vector>
#include <iostream>
#include <string>
#include <map>
#include <stack>
#include <algorithm>
#include <numeric>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <functional>
#include <sstream>
#include <set>

using namespace std;

class Solution
{
public:
    int minSubArrayLen(int target, vector<int> &nums)
    {
        int n = nums.size();
        int l = 0, r = 0, sum = nums[0], res = INT_MAX;

        nums.insert(nums.end(), 0);
        while (r < n)
            if (sum < target)
                sum += nums[++r];
            else
            {
                res = min(res, r - l + 1);
                sum -= nums[l++];
            }

        return res == INT_MAX ? 0 : res;
    }
};
